src/parser/core/modules/EntityStatuses.ts

```diff
@@ -0,0 +1,142 @@
+	static debug = false
+
+	@dependency private invuln!: Invulnerability
+	@dependency private fflogsEvents!: FFLogsEventNormaliser
+
+	// -----
+	// API
+	// -----
+	// TODO: This implementation which I've shamelessly stolen seems to only track overall
+	//       uptime, ignoring potential gains from multidotting, etc. Should that be in here,
+	//       or is that a somewhere-else sort of thing...
+	getStatusUptime(statusId: number, onTargets: TODO, sourceId = this.parser.player.id) {
+		// Search for status maps on valid targets
+		const statusEvents: BuffTrackingEvent[] = Object.values(onTargets)
+			.reduce((statusEvents: BuffTrackingEvent[], target: TODO) => {
+				const appliedStatuses = target.buffs?.filter((statusEvent: BuffTrackingEvent) => {
+					return statusEvent.ability.guid === statusId && (statusEvent.sourceID === null || statusEvent.sourceID === sourceId)
+				})
+				return statusEvents.concat(appliedStatuses)
+			}, [])
+
+		let active = 0
+		let start: number | null = null
+		return statusEvents
+			.reduce((statusRanges: StatusRangeEndpoint[], statusEvent) => {
+				return statusRanges.concat(this.getStatusRangesForEvent(statusEvent))
+			}, [])
+			.sort((a, b) => a.timestamp - b.timestamp)
+			.reduce((uptime, rangeEndpoint) => {
+				this.debug(`Status change ${rangeEndpoint.type} at time ${this.parser.formatTimestamp(rangeEndpoint.timestamp, 1)}`)
+				if (rangeEndpoint.type === APPLY) {
+					if (active === 0) {
+						this.debug('Status not currently active on any targets, starting new window')
+						start = rangeEndpoint.timestamp
+					}
+					active++
+				}
+				if (rangeEndpoint.type === REMOVE) {
+					active--
+					if (active === 0) {
+						uptime += rangeEndpoint.timestamp - (start ?? rangeEndpoint.timestamp)
+						this.debug(`Status ended on all targets.  Total uptime now ${this.parser.formatDuration(uptime, 1)}`)
+					}
+				}
+				return uptime
+			}, 0)
+	}
+
+	getStatusRangesForEvent(statusEvent: BuffTrackingEvent) {
+		this.debug(`Determining active time range for status ${statusEvent.ability.name} - started at: ${this.parser.formatTimestamp(statusEvent.start, 1)} - ended at: ${(statusEvent.end)? this.parser.formatTimestamp(statusEvent.end, 1) : ''}`)
+		return this.splitStatusEventForInvulns(statusEvent).reduce((statusRanges: StatusRangeEndpoint[], statusEvent) => {
+			statusRanges.push({timestamp: statusEvent.start, type: APPLY, stackHistory: statusEvent.stackHistory})
+			statusRanges.push({timestamp: statusEvent.end ?? this.parser.currentTimestamp, type: REMOVE, stackHistory: statusEvent.stackHistory})
+			return statusRanges
+		}, [])
+	}
+
+	splitStatusEventForInvulns(statusEvent: BuffTrackingEvent) {
+		const eventToAdjust = {...statusEvent}
+		const adjustedEvents = [eventToAdjust]
+		if (!statusEvent.end) {
+			this.debug('Unfinished status event detected.  Applying ability duration after last refresh event.')
+			const statusInfo = getDataBy(STATUSES, 'id', statusEvent.ability.guid)
```

Comment:
Esp. given this is core, `this.data.getStatus(statusEvent.ability.guid)` - you'll need to add data as a dep (should be safe, `data` is and should remain completely depless).